server/tilesource/geospatial.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+#############################################################################
+#  Copyright Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#############################################################################
+
+import gdal
+import osr
+import mapnik
+from PIL import Image
+import math
+
+from .base import FileTileSource
+
+try:
+    import girder
+    from girder import logger
+    from .base import GirderTileSource
+except ImportError:
+    girder = None
+    import logging as logger
+    logger.getLogger().setLevel(logger.INFO)
+
+
+class MapnikTileSource(FileTileSource):
+    name = 'mapnik'
+
+    def __init__(self, path, **kwargs):
+        super(MapnikTileSource, self).__init__(path, **kwargs)
+        self.dataset = gdal.Open(self._getLargeImagePath())
+
+        self.sizeX = self.dataset.RasterXSize
+        self.sizeY = self.dataset.RasterYSize
+        self.tileWidth = 256
+        self.tileHeight = 256
+        # # Fair assumption
+        self.levels = 20
+
+        # Earth circumference for web mercator
+        self.circumference = 20037508.342789244
+
+    def _getProj4String(self):
+        """Returns proj4 string for the given dataset"""
+        wkt = self.dataset.GetProjection()
+        proj = osr.SpatialReference()
+        proj.ImportFromWkt(wkt)
+
+        return proj.ExportToProj4()
+
+    def getPixelSize(self, zoom):
+        """Returns an approximate pixel size for a given zoom level"""
+
+        equatorLatitude = 0
+        # For simplification we assume everything is equator latitude
+        pixelSize = self.circumference * math.cos(equatorLatitude) / 2 ** (zoom + 7)
```

Comment:
To be clear about why the division occurs, this should probably be `pixelSize = self.circumference * math.cos(equatorLatitude) / self.pixelWidth`.

Reply:
http://wiki.openstreetmap.org/wiki/Zoom_levels I used the formula from here. Does that make sense?

Reply:
I mistyped my replacement, which should have been `pixelSize = self.circumference * math.cos(equatorLatitude) / 2 ** zoom / self.pixelWidth`.  The `+ 8` is just because the tile width happens to be `2 ** 8`, but, there is no intrinsic requirement that the tile size be a power of two, so making it explicit where this factor comes from is clearer.